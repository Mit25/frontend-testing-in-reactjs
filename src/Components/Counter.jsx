import React from 'react';

export default class Counter extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            count: 0
        }
    }

    render() {
        return (
            <div>
                <div>
                    <span id='spn-counter-text'>The counter: </span>
                    <span id='spn-count'>{this.state.count}</span>
                </div>
                <button id='btn-increment' onClick={e => this.increment(e)}>Increment</button>
                <button id='btn-decrement' onClick={e => this.decrement(e)}>Decrement</button>
                <button id='btn-reset' onClick={e => this.reset(e)}>Reset</button>
            </div>
        )
    }

    makeIncrementer(amount, event) {
        this.setState({
            count: this.state.count + amount
        });
    }

    reset(event) {
        this.setState({
            count: 0
        });
    }

    increment = (event) => this.makeIncrementer(1, event);
    decrement = (event) => this.makeIncrementer(-1, event);

}

Counter.defaultProps = {

};