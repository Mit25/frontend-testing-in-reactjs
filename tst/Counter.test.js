import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Counter from '../src/Components/Counter';

describe('Counter component', () => {

    it('snapshot testing for Counter component', () => {
        const tree = renderer.create(<Counter />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render Counter component with initial counter', () => {
        const wrapper = shallow(<Counter />);

        expect(wrapper.find('#spn-count').text()).toBe('0');
    });

    describe('testing button functionality', () => {

        it('should increment counter on Increment button click asssert via state', () => {
            const wrapper = shallow(<Counter />);
            wrapper.find('#btn-increment').simulate('click');

            /*
             * This is not good way to test.
             *
             * Since we only assert internal state value that does not guarantee that it will display correct thing
             * on webpage.
             */
            expect(wrapper.state('count')).toBe(1);
        });

        it('should increment counter on Increment button click', () => {
            const wrapper = shallow(<Counter />);
            wrapper.find('#btn-increment').simulate('click');

            expect(wrapper.find('#spn-count').text()).toBe('1');
        });

        it('should decrement counter on Decrement button click', () => {
            const wrapper = shallow(<Counter />);
            wrapper.find('#btn-decrement').simulate('click');

            expect(wrapper.find('#spn-count').text()).toBe('-1');
        });

        it('should reset counter on Reset button click', () => {
            const wrapper = shallow(<Counter />);

            wrapper.find('#btn-increment').simulate('click');
            expect(wrapper.find('#spn-count').text()).toBe('1');

            wrapper.find('#btn-reset').simulate('click');
            expect(wrapper.find('#spn-count').text()).toBe('0');
        });
    });
});