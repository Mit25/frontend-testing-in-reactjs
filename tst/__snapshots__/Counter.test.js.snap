// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component snapshot testing for Counter component 1`] = `
<div>
  <div>
    <span
      id="spn-counter-text"
    >
      The counter: 
    </span>
    <span
      id="spn-count"
    >
      0
    </span>
  </div>
  <button
    id="btn-increment"
    onClick={[Function]}
  >
    Increment
  </button>
  <button
    id="btn-decrement"
    onClick={[Function]}
  >
    Decrement
  </button>
  <button
    id="btn-reset"
    onClick={[Function]}
  >
    Reset
  </button>
</div>
`;
